crates/block-multiplier/src/lib.rs

```diff
@@ -35,3 +43,468 @@ macro_rules! subarray{
+    let mut s0_t = [0_u64; 8];
+    let mut carry = 0;
+    (s0_t[0], carry) = carrying_mul_add(s0_a[0], s0_b[0], s0_t[0], carry);
+    (s0_t[1], carry) = carrying_mul_add(s0_a[0], s0_b[1], s0_t[1], carry);
+    (s0_t[2], carry) = carrying_mul_add(s0_a[0], s0_b[2], s0_t[2], carry);
+    (s0_t[3], carry) = carrying_mul_add(s0_a[0], s0_b[3], s0_t[3], carry);
+    s0_t[4] = carry;
+    carry = 0;
+    (s0_t[1], carry) = carrying_mul_add(s0_a[1], s0_b[0], s0_t[1], carry);
+    (s0_t[2], carry) = carrying_mul_add(s0_a[1], s0_b[1], s0_t[2], carry);
+    (s0_t[3], carry) = carrying_mul_add(s0_a[1], s0_b[2], s0_t[3], carry);
+    (s0_t[4], carry) = carrying_mul_add(s0_a[1], s0_b[3], s0_t[4], carry);
+    s0_t[5] = carry;
+    carry = 0;
+    (s0_t[2], carry) = carrying_mul_add(s0_a[2], s0_b[0], s0_t[2], carry);
+    (s0_t[3], carry) = carrying_mul_add(s0_a[2], s0_b[1], s0_t[3], carry);
+    (s0_t[4], carry) = carrying_mul_add(s0_a[2], s0_b[2], s0_t[4], carry);
+    (s0_t[5], carry) = carrying_mul_add(s0_a[2], s0_b[3], s0_t[5], carry);
+    s0_t[6] = carry;
+    carry = 0;
+    (s0_t[3], carry) = carrying_mul_add(s0_a[3], s0_b[0], s0_t[3], carry);
+    (s0_t[4], carry) = carrying_mul_add(s0_a[3], s0_b[1], s0_t[4], carry);
+    (s0_t[5], carry) = carrying_mul_add(s0_a[3], s0_b[2], s0_t[5], carry);
+    (s0_t[6], carry) = carrying_mul_add(s0_a[3], s0_b[3], s0_t[6], carry);
+    s0_t[7] = carry;
+
+    let mut s0_r1 = [0_u64; 5];
+    (s0_r1[0], s0_r1[1]) = carrying_mul_add(s0_t[0], U64_I3[0], s0_r1[0], 0);
+    (s0_r1[1], s0_r1[2]) = carrying_mul_add(s0_t[0], U64_I3[1], s0_r1[1], 0);
+    (s0_r1[2], s0_r1[3]) = carrying_mul_add(s0_t[0], U64_I3[2], s0_r1[2], 0);
+    (s0_r1[3], s0_r1[4]) = carrying_mul_add(s0_t[0], U64_I3[3], s0_r1[3], 0);
+
+    let mut s0_r2 = [0_u64; 5];
+    (s0_r2[0], s0_r2[1]) = carrying_mul_add(s0_t[1], U64_I2[0], s0_r2[0], 0);
+    (s0_r2[1], s0_r2[2]) = carrying_mul_add(s0_t[1], U64_I2[1], s0_r2[1], 0);
+    (s0_r2[2], s0_r2[3]) = carrying_mul_add(s0_t[1], U64_I2[2], s0_r2[2], 0);
+    (s0_r2[3], s0_r2[4]) = carrying_mul_add(s0_t[1], U64_I2[3], s0_r2[3], 0);
+
+    let mut s0_r3 = [0_u64; 5];
+    (s0_r3[0], s0_r3[1]) = carrying_mul_add(s0_t[2], U64_I1[0], s0_r3[0], 0);
+    (s0_r3[1], s0_r3[2]) = carrying_mul_add(s0_t[2], U64_I1[1], s0_r3[1], 0);
+    (s0_r3[2], s0_r3[3]) = carrying_mul_add(s0_t[2], U64_I1[2], s0_r3[2], 0);
+    (s0_r3[3], s0_r3[4]) = carrying_mul_add(s0_t[2], U64_I1[3], s0_r3[3], 0);
+
+    let s0_s = addv(addv(subarray!(s0_t, 3, 5), s0_r1), addv(s0_r2, s0_r3));
+
+    let s0_m = U64_MU0.wrapping_mul(s0_s[0]);
+    let mut s0_mp = [0_u64; 5];
+    (s0_mp[0], s0_mp[1]) = carrying_mul_add(s0_m, P[0], s0_mp[0], 0);
+    (s0_mp[1], s0_mp[2]) = carrying_mul_add(s0_m, P[1], s0_mp[1], 0);
+    (s0_mp[2], s0_mp[3]) = carrying_mul_add(s0_m, P[2], s0_mp[2], 0);
+    (s0_mp[3], s0_mp[4]) = carrying_mul_add(s0_m, P[3], s0_mp[3], 0);
+
+    let s0 = subarray!(addv(s0_s, s0_mp), 1, 4);
```

Comment:
`addv` is not necessary. Those addition can be pulled into `carrying_mul_add`. That's what I do in the assembly implementation.